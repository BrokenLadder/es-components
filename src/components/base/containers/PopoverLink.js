import React from 'react';
import PropTypes from 'prop-types';
import Overlay from 'react-overlays/lib/Overlay';
import { noop } from 'lodash';
import Button from '../../controls/buttons/Button';
import Popover from './Popover';
import FocusTrap from 'focus-trap-react';

function getArrowPlacement(popoverPlacement) {
  switch (popoverPlacement) {
    case 'top':
    case 'bottom':
      return popoverPlacement;
    case 'right':
      return 'left';
    case 'left':
      return 'right';
    default:
      return popoverPlacement;
  }
}

/*
  monkey-patching the FocusTrap component
  fixes a timing issue with the popover
*/
FocusTrap.prototype.componentDidMount = function() {
  const specifiedFocusTrapOptions = this.props.focusTrapOptions;
  const tailoredFocusTrapOptions = {
    returnFocusOnDeactivate: false
  };
  for (const optionName in specifiedFocusTrapOptions) {
    if (!specifiedFocusTrapOptions.hasOwnProperty(optionName)) continue;
    if (optionName === 'returnFocusOnDeactivate') continue;
    tailoredFocusTrapOptions[optionName] =
      specifiedFocusTrapOptions[optionName];
  }

  this.focusTrap = this.props._createFocusTrap(
    this.node,
    tailoredFocusTrapOptions
  );
  if (this.props.active) {
    setTimeout(() => {
      this.focusTrap.activate();
    }, 0);
  }
  if (this.props.paused) {
    this.focusTrap.pause();
  }
}

class PopoverLink extends React.Component {
  static propTypes = {
    linkContent: PropTypes.node.isRequired,
    popoverTitle: PropTypes.string,
    /** The content displayed in the popover body */
    popoverContent: PropTypes.node.isRequired,
    /** The placement of the popover in relation to the link */
    popoverPlacement: PropTypes.oneOf(['top', 'right', 'bottom', 'left']),
    containsFormElement: PropTypes.bool,
    /** Function to run when the popover is hidden */
    onPopoverHidden: PropTypes.func
  }

  static defaultProps = {
    popoverPlacement: 'top',
    containsFormElement: false,
    onPopoverHidden: noop
  }

  constructor(props) {
    super(props);

    this.state = { isOpen: false };
  }

  showPopover = () => {
    this.setState({ isOpen: true });
  }

  hidePopover = () => {
    if(this.state.isOpen) {
      this.setState({ isOpen: false });
      this.props.onPopoverHidden();
    }
  }

  render() {
    const {
      linkContent,
      popoverTitle,
      popoverContent,
      popoverPlacement,
      containsFormElement
    } = this.props;

    const { isOpen } = this.state;
    const arrowPlacement = getArrowPlacement(popoverPlacement);
    const shouldCloseOnAnyClick = !containsFormElement;
    const focusTrapOptions = {
      onDeactivate: this.hidePopover
    };

    return (
      <span>
        <span ref={(span) => { this.popoverTarget = span; }}>
          <Button data-trigger="focus"
            styleType="link"
            handleOnClick={this.showPopover}
          >
            {linkContent}
          </Button>
        </span>

        <Overlay
          show={isOpen}
          placement={popoverPlacement}
          target={this.popoverTarget}
          rootClose={shouldCloseOnAnyClick}
          onHide={this.hidePopover}
        >
          <Popover title={popoverTitle} arrowPlacement={arrowPlacement} containsFormElement={containsFormElement}>
            <FocusTrap active={containsFormElement} focusTrapOptions={focusTrapOptions}>
              {popoverContent}
            </FocusTrap>
          </Popover>
        </Overlay>
      </span>
    );
  }
}

export default PopoverLink;
