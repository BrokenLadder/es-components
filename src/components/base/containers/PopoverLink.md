```
<p>
  Click
  <PopoverLink
    linkContent="here "
    popoverTitle="Popover"
    popoverContent="This is the popover's content. lkalakjdflkaj flakdjf lakdjf lakdfjl adkfj akdfjalkd fjalkdfj ladkfj kajdffj adf adkflja ldfjlad kfjadlkfj."
  /> to activate the popover.
</p>
```

```
function PopoverForm() {
  return (
    <div>
      <Textbox
        id="firstName"
        labelText="First name"
      />
      <Textbox
        labelText="Last name"
      />

      <Button handleOnClick={() => {}}>
        Submit
      </Button>
    </div>
  );
}

const popoverContent = <PopoverForm />;

<p>
  You can even put forms inside of a
  <PopoverLink
    linkContent="popover"
    popoverContent={popoverContent}
    containsFormElement
  />.
</p>
```

```
const popoverContent = <Icon name="question-sign" />;

<p>
  Click the question mark for some help text.
  <PopoverLink
    linkContent={popoverContent}
    popoverContent="Here's some help text"
  />
</p>
```
