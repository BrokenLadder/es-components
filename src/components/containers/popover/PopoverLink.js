import React from 'react';
import PropTypes from 'prop-types';
import { noop } from 'lodash';
import styled, { ThemeProvider } from 'styled-components';
import uncontrollable from 'uncontrollable';

import { buttonStyleTypes } from '../../controls/buttons/button-variants';
import defaultTheme from '../../theme/defaultTheme';
import Button from '../../controls/buttons/Button';
import PopoverTrigger from './PopoverTrigger';

const PopoverButton = styled(Button)`
  border-bottom: ${props => (props.suppressUnderline ? 'none' : '1px dotted')};
  text-decoration: none;
`;

export class PopoverLink extends React.Component {
  toggleShow = () => {
    const showPopover = !this.props.showPopover;
    this.props.onToggle(showPopover);
  };

  hidePopover = () => {
    if (this.props.showPopover) {
      this.toggleShow();
    }
  };

  render() {
    const {
      children,
      popoverTitle,
      popoverContent,
      popoverPlacement,
      showPopover,
      showCloseButton,
      suppressUnderline,
      theme,
      buttonStyle
    } = this.props;

    const styledLink = buttonStyle !== 'link';

    return (
      <ThemeProvider theme={theme}>
        <PopoverTrigger
          onHideOverlay={this.hidePopover}
          popoverContent={popoverContent}
          popoverPlacement={popoverPlacement}
          popoverTarget={this.popoverTarget}
          popoverTitle={popoverTitle}
          showPopover={showPopover}
          showCloseButton={showCloseButton}
        >
          <PopoverButton
            aria-haspopup="dialog"
            data-trigger="focus"
            handleOnClick={this.toggleShow}
            ref={btn => {
              this.popoverTarget = btn;
            }}
            styleType={buttonStyle}
            styledLink={styledLink}
            suppressUnderline={suppressUnderline}
            theme={theme}
          >
            {children}
          </PopoverButton>
        </PopoverTrigger>
      </ThemeProvider>
    );
  }
}

PopoverLink.propTypes = {
  /** The link content which activates the popover */
  children: PropTypes.node.isRequired,
  /** The text displayed in the popover title section */
  popoverTitle: PropTypes.string,
  /** The content displayed in the popover body */
  popoverContent: PropTypes.node.isRequired,
  /** The placement of the popover in relation to the link */
  popoverPlacement: PropTypes.oneOf(['top', 'right', 'bottom', 'left']),
  /** Function to toggle display of popover */
  onToggle: PropTypes.func,
  /** Displays or hides the popover */
  showPopover: PropTypes.bool,
  /** Display a close ('x') button */
  showCloseButton: PropTypes.bool,
  /** Hide underline from link. Useful for children like Icons */
  suppressUnderline: PropTypes.bool,
  /**
   * Theme object used by the ThemeProvider,
   * automatically passed by any parent component using a ThemeProvider
   */
  theme: PropTypes.object,
  /** The style of the link for the popover */
  buttonStyle: PropTypes.oneOf(buttonStyleTypes)
};

PopoverLink.defaultProps = {
  popoverPlacement: 'top',
  onToggle: noop,
  showPopover: false,
  showCloseButton: false,
  suppressUnderline: false,
  theme: defaultTheme,
  buttonStyle: 'link'
};

const UncontrolledPopoverLink = uncontrollable(PopoverLink, {
  showPopover: 'onToggle'
});

export default UncontrolledPopoverLink;
