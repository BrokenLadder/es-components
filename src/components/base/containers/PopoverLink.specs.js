/* eslint-env jest */

import React from 'react';
import renderer from 'react-test-renderer';

import PopoverLink from './PopoverLink';

describe('popoverLink component', () => {
  it('renders popoverLink children', () => {
    const tree = renderer.create(
      <PopoverLink popoverTitle="Popover Title" popovercontent="This is a popover.">
        This is the popover link text.
      </PopoverLink>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
