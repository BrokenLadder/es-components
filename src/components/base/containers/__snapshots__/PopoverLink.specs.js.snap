exports[`popoverLink component renders popoverLink children 1`] = `
<span>
  <span>
    <button
      className="kKThPK"
      data-trigger="focus"
      onClick={[Function]}>
      This is the popover link text.
    </button>
  </span>
</span>
`;
